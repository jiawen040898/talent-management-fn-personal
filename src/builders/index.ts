export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
