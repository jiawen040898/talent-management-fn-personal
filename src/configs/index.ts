export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
