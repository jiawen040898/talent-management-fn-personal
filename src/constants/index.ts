export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
