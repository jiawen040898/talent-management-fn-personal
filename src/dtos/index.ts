export * from './company.dto';
export * from './cron-schedule-message.dto';
export * from './employee.dto';
export * from './employee-account-created-event.dto';
export * from './employee-assessment-score-event.dto';
export * from './employee-fit-score-event.dto';
export * from './feedback-cycle.dto';
export * from './fit-score-recipe.dto';
export * from './manager-subordinate-aggregated-report-data.dto';
