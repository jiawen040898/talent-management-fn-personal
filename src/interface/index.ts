export * from '../dtos/fit-score-recipe.dto';
export * from './cognitive-score.interface';
export * from './employee-score.interface';
export * from './event-model.interface';
export * from './feedback-cycle.interface';
export * from './feedback-recommendation.interface';
export * from './fit-score.interface';
export * from './personality-score.interface';
export * from './program.interface';
export * from './redis.interface';
