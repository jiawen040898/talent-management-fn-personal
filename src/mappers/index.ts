export * from './employee-score.mapper';
export * from './feedback.mapper';
export * from './feedback-recommendation.mapper';
