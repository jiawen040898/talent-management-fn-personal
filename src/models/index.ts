export * from './employee.entity';
export * from './employee-questionnaire.entity';
export * from './employee-score.entity';
export * from './employee-task.entity';
export * from './entities';
export * from './feedback-cycle.entity';
export * from './feedback-cycle-email-template.entity';
export * from './feedback-cycle-notification.entity';
export * from './feedback-cycle-reviewee.entity';
export * from './feedback-cycle-reviewer.entity';
export * from './learning-goal.entity';
export * from './participant.entity';
export * from './program.entity';
