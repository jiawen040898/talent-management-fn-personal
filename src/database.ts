import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
