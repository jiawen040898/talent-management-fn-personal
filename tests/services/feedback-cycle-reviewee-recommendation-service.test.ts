import { snsService } from '@pulsifi/fn';
import { SQSEvent } from 'aws-lambda';
import { DataSource } from 'typeorm';

import * as database from '../../src/database';
import { FeedbackCycleReviewee } from '../../src/models';
import { FeedbackCycleRevieweeRecommendation } from '../../src/models/feedback-cycle-reviewee-recommendation.entity';
import { feedbackCycleRevieweeRecommendationService } from '../../src/services/feedback-cycle-reviewee-recommendation.service';
import { getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/test-data';

jest.mock('@pulsifi/fn/services/sns.service');

jest.mock('../../src/services/recommendation.service', () => {
    const mBInstance = {
        getActionWithSkillsRecommendation: jest.fn(
            () => testData.mockRecommendationActionResponse,
        ),
        getResourcesRecommendation: jest.fn(
            () => testData.mockRecommendationResourceResponse,
        ),
    };
    const mB = jest.fn(() => mBInstance);
    return { RecommendationService: mB };
});

jest.mock('../../src/services/company.service', () => {
    const mBInstance = {
        getCompanyById: jest.fn(() => testData.mockCompanyLookupData),
    };
    const mB = jest.fn(() => mBInstance);
    return { CompanyService: mB };
});

describe('feedbackCycleRevieweeRecommendationService', () => {
    let dataSource: DataSource;
    const spygetDataSource = jest.spyOn(database, 'getDataSource');
    const urlMatchedRegex = '^https://local-employee.pulsifi.me/';

    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesFeedbackCycleRecommendationToBeAdded,
        );
        spygetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('handleRecommendationProcessRequest', async () => {
        // Act
        const results =
            await feedbackCycleRevieweeRecommendationService.handleRecommendationProcessRequest(
                testData.mockFeedbackRecommendationGenerationProcessEvent as SQSEvent,
            );

        //Assert
        expect({ results }).toMatchSnapshot({
            results: new Array(results.length).fill({
                id: expect.anything(),
                feedback_cycle_reviewee_id: expect.anything(),
            }),
        });
    });

    it('handleRecommendationGenerationRequest', async () => {
        // Act
        const results =
            await feedbackCycleRevieweeRecommendationService.handleRecommendationGenerationRequest(
                testData.mockFeedbackRecommendationGenerationRequestData,
            );

        //Assert
        expect({ results }).toMatchSnapshot({
            results: {
                id: expect.anything(),
                feedback_cycle_reviewee_id: expect.anything(),
            },
        });
        const reviewee = await dataSource
            .getRepository(FeedbackCycleReviewee)
            .findOne({
                select: {
                    id: true,
                },
                where: {
                    id: testData.reviewee1.id,
                },
            });

        expect(reviewee).toMatchSnapshot({
            id: expect.anything(),
        });

        const revieweeRecommendation = await dataSource
            .getRepository(FeedbackCycleRevieweeRecommendation)
            .find({
                select: {
                    id: true,
                    skill: true,
                    action: true,
                    resource_course: true,
                    resource_book: true,
                },
                where: {
                    feedback_cycle_reviewee_id: testData.reviewee1.id,
                },
            });

        expect({ results: revieweeRecommendation }).toMatchSnapshot({
            results: new Array(revieweeRecommendation.length).fill({
                id: expect.anything(),
                action: new Array(
                    revieweeRecommendation[0].action?.length,
                ).fill({
                    id: expect.anything(),
                }),
                resource_book: new Array(
                    revieweeRecommendation[0].resource_book?.length,
                ).fill({
                    id: expect.anything(),
                }),
                resource_course: new Array(
                    revieweeRecommendation[0].resource_course?.length,
                ).fill({
                    id: expect.anything(),
                }),
            }),
        });
    });

    it('handleRecommendationEndedRequest', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();

        // Act
        await feedbackCycleRevieweeRecommendationService.handleRecommendationEndedRequest(
            testData.mockFeedbackRecommendationGenerationEndedEvent,
        );

        // Assert
        const reviewee = await dataSource
            .getRepository(FeedbackCycleReviewee)
            .findOne({
                select: ['id', 'recommendation_result_meta'],
                where: {
                    id: testData.reviewee6.id,
                },
            });

        expect(reviewee).toMatchSnapshot({
            id: expect.anything(),
            recommendation_result_meta: {
                generated_at: expect.any(String),
            },
        });

        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                receiver: {
                    id: expect.anything(),
                    employee_id: expect.anything(),
                },
                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });
    });
});
