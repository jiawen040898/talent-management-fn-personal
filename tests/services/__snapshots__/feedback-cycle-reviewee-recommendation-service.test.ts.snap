// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationEndedRequest 1`] = `
{
  "id": Anything,
  "recommendation_result_meta": {
    "generated_at": Any<String>,
    "status": "ready",
  },
}
`;

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationEndedRequest 2`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "receiver": {
      "email": "employee0@test.pulsifi.me",
      "employee_id": Anything,
      "first_name": "First Name 0",
      "full_name": "First Name 0 Last Name 0",
      "id": Anything,
      "last_name": "Last Name 0",
      "user_account_id": 100,
    },
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_reviewee_recommendation_completed",
  "user_account_id": 100,
}
`;

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationGenerationRequest 1`] = `
{
  "results": {
    "feedback_cycle_reviewee_id": Anything,
    "id": Anything,
  },
}
`;

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationGenerationRequest 2`] = `
{
  "id": Anything,
}
`;

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationGenerationRequest 3`] = `
{
  "results": [
    {
      "action": [
        {
          "description": "some explanation",
          "id": Anything,
          "title": "some action",
        },
      ],
      "id": Anything,
      "resource_book": [
        {
          "author": [
            "some author",
          ],
          "id": Anything,
          "image_url": "www.image.url",
          "redirect_url": "www.info.url",
          "title": "some title",
        },
      ],
      "resource_course": [
        {
          "author": [
            "some author",
          ],
          "id": Anything,
          "image_url": "www.image.url",
          "redirect_url": "www.info.url",
          "title": "some title",
        },
      ],
      "skill": null,
    },
  ],
}
`;

exports[`feedbackCycleRevieweeRecommendationService handleRecommendationProcessRequest 1`] = `
{
  "results": [
    {
      "feedback_cycle_reviewee_id": Anything,
      "id": Anything,
      "question_description": "Generates creative ideas and alternative solutions to overcome challenges.",
      "question_title": "Critical Thinking",
      "recommendation_dimension": "development",
      "score": "5.0000000",
    },
  ],
}
`;
